[dynamo] Organize higherorderops variable trackers

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104639
* #104577
* __->__ #104565

The main change is moving the higherorderops from torch.py to higher_order_ops.py. And creating smaller subclasses of HigherOrderOp for cond, map etc

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

anijain2305 committed:
[dynamo] Organize higherorderops variable trackers

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Organize higherorderops variable trackers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Organize higherorderops variable trackers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Organize higherorderops variable trackers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Organize higherorderops variable trackers"


The main change is moving the higherorderops from torch.py to higher_order_ops.py. And creating smaller subclasses of HigherOrderOp for cond, map etc

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -0,0 +1,916 @@
+        )
+
+
+class HigherOrderWrapVariable(TorchHigherOrderOperatorVariable):
```
There's a lot of code duplication here:
- Everyone needs to check that "only constant kwargs are supported" -> this should probably be a helper function
- HigherOrderWrapVariable and HigherOrderCheckpointVariable do the same thing. HigherOrderCheckpointVariable and HigherOrderWrapVariable should probably inherit from a common base that has the call_function. There will be other HigherOrderOps in the future that behave similarly.

In general, we should try to reuse code among these implementations as much as possible so that future improvements (e.g. support for body_functions that accept pytrees) can easily be added across the higher order operators.

zou3519 reviewed (commented):
I'm assuming you just refactored the code. I had some suggestions for code reuse. Otherwise, this looks good to me and it is awesome that we are refactoring this code to be more maintainable

anijain2305 committed:
Update on "[dynamo] Organize higherorderops variable trackers"


The main change is moving the higherorderops from torch.py to higher_order_ops.py. And creating smaller subclasses of HigherOrderOp for cond, map etc

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

anijain2305 commented:
@zou3519 I incorporated all your comments.

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -0,0 +1,848 @@
+        )
+
+
+class HigherOrderFuncTorchGradVariable(TorchHigherOrderOperatorVariable):
```
nit: FuncTorchGradHigherOrderVariable

anijain2305 replied:
Ah, missed this one. Thanks for catching it!

zou3519 reviewed (approved):
LGTM, one last nit

anijain2305 committed:
Update on "[dynamo] Organize higherorderops variable trackers"


The main change is moving the higherorderops from torch.py to higher_order_ops.py. And creating smaller subclasses of HigherOrderOp for cond, map etc

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge

voznesenskym commented:
posthumous LGTM :P
